lib/web_ui/lib/src/engine/layers.dart

```diff
@@ -531,15 +531,211 @@ class PlatformViewStyling {
     if (other is! PlatformViewStyling) {
       return false;
     }
-    return (position == other.position) && (opacity == other.opacity);
+    return (position == other.position) && (opacity == other.opacity) && (clip == other.clip);
   }
 
   @override
   int get hashCode {
-    return Object.hash(position, opacity);
+    return Object.hash(position, opacity, clip);
   }
 }
 
+sealed class PlatformViewClip {
+  PlatformViewClip positioned(PlatformViewPosition position);
+
+  ui.Rect get boundingRect;
+
+  bool covers(ui.Rect rect);
+
+  ScenePath get asPath;
+
+  static PlatformViewClip combine(PlatformViewClip outer, PlatformViewClip inner) {
+    if (outer is PlatformViewNoClip) {
+      return inner;
+    }
+    if (inner is PlatformViewNoClip) {
+      return outer;
+    }
+
+    if (outer.covers(inner.boundingRect)) {
+      return outer;
+    }
+
+    if (inner.covers(outer.boundingRect)) {
+      return inner;
+    }
+
+    final ScenePath path = ui.Path() as ScenePath;
+    path.addPath(outer.asPath, ui.Offset.zero);
+    path.addPath(inner.asPath, ui.Offset.zero);
+    return PlatformViewPathClip(path);
+  }
+}
+
+class PlatformViewNoClip implements PlatformViewClip {
+  const PlatformViewNoClip();
+
+  @override
+  PlatformViewClip positioned(PlatformViewPosition positioned) {
+    return this;
+  }
+
+  @override
+  ScenePath get asPath => ui.Path() as ScenePath;
+
+  @override
+  ui.Rect get boundingRect => ui.Rect.zero;
+
+  @override
+  bool covers(ui.Rect rect) {
+    return false;
+  }
+
+  @override
+  bool operator ==(Object other) {
+    if (identical(this, other) || (other.runtimeType == runtimeType)) {
+      return true;
+    }
+    return false;
+  }
+
+  @override
+  int get hashCode => runtimeType.hashCode;
+}
+
+class PlatformViewRectClip implements PlatformViewClip {
+  PlatformViewRectClip(this.rect);
+
+  final ui.Rect rect;
+
+  @override
+  PlatformViewClip positioned(PlatformViewPosition position) {
+    if (position.isZero) {
+      return this;
+    }
+    final ui.Offset? offset = position.offset;
+    if (offset != null) {
+      return PlatformViewRectClip(rect.shift(offset));
+    } else {
+      return PlatformViewPathClip(asPath.transform(position.transform!.toFloat64()) as ScenePath);
+    }
+  }
+
+  @override
+  ScenePath get asPath => (ui.Path() as ScenePath)..addRect(rect);
+
+  @override
+  ui.Rect get boundingRect => rect;
+
+  @override
+  bool covers(ui.Rect other) {
+    return rect.left <= other.left &&
+      rect.right >= other.right &&
+      rect.top <= other.top &&
+      rect.bottom >= other.bottom;
+  }
+
+  @override
+  bool operator ==(Object other) {
+    if (other.runtimeType != runtimeType) {
+      return false;
+    }
+    return other is PlatformViewRectClip && rect == other.rect;
+  }
+
+  @override
+  int get hashCode => Object.hash(runtimeType, rect);
```

Comment:
Well we want to avoid hash collisions in cases where they are unequal to avoid the comparison in the first place (they will be bucketed separately). I looked through our code and I found several other examples where polymorphic objects that implement `hashCode` do so by including the `runtimeType` in the hash, which seems correct and sane to me. But again, this code will probably never actually be executed and probably will just be tree-shaken out, since these are never put in a hashed collection anyway. This implementation is just to satisfy the lint.